### PR TITLE
8258147: Modernize Nashorn code

### PR DESCRIPTION
**Pull request by szegedi**
Most of Nashorn code was written to be compatible with Java 7 and never updated. It is high time to introduce diamonds, lambdas, modern collection operations (`List.of`, `Map.computeIfAbsent`, etc.), elimination of most StringBuffer use (since string concatenation has been indified) and bring it up to at least Java 11 standards.

In addition to these, I added more code tidying mostly governed by judicious evaluation of IntelliJ IDEA recommendations. Expressions have been simplified, statically known values have been respected, unused parameters and dead code removed. I did not blindly follow IDEA's recommendations; I refrained from eliminating asserts even when they're statically proven to always be true, as I like them as safeguards against future breakage. I also did not convert any switch statements into switch expressions; those have been introduced very recently in Java 14 and I'd rather give them more time. Who knows, maybe some poor soul out there plans to compile standalone Nashorn for Java 11?

All tests, including the 262 test suite, pass after these changes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258147](https://bugs.openjdk.java.net/browse/JDK-8258147): Modernize Nashorn code

### Reviewers
 * [Athijegannathan Sundararajan](https://openjdk.java.net/census#sundar) (@sundararajana - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/nashorn pull/5/head:pull/5`
`$ git checkout pull/5`

### CODE REVIEW

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/runtime/Context.java [hunk 1]
```diff
@@ -829,7 +829,7 @@ public Object eval(final ScriptObject initialScope, final String string,
 
         Class<?> clazz;
         try {
-            clazz = compile(source, new ThrowErrorManager(), strictFlag, true);
+            clazz = compile(source, new ThrowErrorManager(), strictFlag);
```
Comment (sundararajana) [inline src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/runtime/Context.java:832]: I vaguely isEval or not was used in compilation to compile differently. Or perhaps only for loading post compilation? (eval code as throw-away anon class?). But if the current code does not use it, then fine.

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/api/scripting/URLReader.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public URLReader(final URL url, final Charset cs) {
     }
 
     @Override
-    public int read(final char cbuf[], final int off, final int len) throws IOException {
+    public int read(final char[] cbuf, final int off, final int len) throws IOException {
```
Comment (sundararajana) [inline src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/api/scripting/URLReader.java:85]: Oops. Never knew we had those in the code :) 

Did you run any tool to find these or just found while code inspection/reading?

**Review by sundararajana — APPROVED** [review]
Thanks for doing this cleanup. Reads much better now. 

I recommend raising many PRs instead of one big PR. I understand this is code cleanup but easier to work smaller changes at a time.

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/api/scripting/URLReader.java (hunk 2 above)
Comment (szegedi) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/api/scripting/URLReader.java:85]: IntelliJ IDEA warns about them. Most of this code changes were driven by IDEA highlighting things.

**Pull request closed without merging**